src/storage/src/source/postgres/snapshot.rs

```diff
@@ -478,3 +571,160 @@ fn decode_copy_row(data: &[u8], col_len: usize, row: &mut Row) -> Result<(), Def
     }
     Ok(())
 }
+
+/// Record the sizes of the tables being snapshotted in `PgSnapshotMetrics`.
+async fn record_table_sizes(
+    config: &RawSourceCreationConfig,
+    connection_config: &Config,
+    snapshot: &str,
+    metrics: PgSnapshotMetrics,
+    // The table names and oids owned by this worker.
+    tables: Vec<(String, Oid)>,
+    // All tables (names and oids) in the snapshot)
+    all_tables: Vec<(String, Oid)>,
+    // If this worker is the snapshot leader.
+    leader: bool,
+) -> Result<Option<AbortOnDropHandle<Result<(), anyhow::Error>>>, anyhow::Error> {
+    let snapshot_config = config.config.parameters.pg_snapshot_config;
+    let statement_timeout = config
+        .config
+        .parameters
+        .pg_source_snapshot_statement_timeout;
+    let connection_context = &config.config.connection_context;
+
+    let source_id = config.id;
+    let worker_id = config.worker_id;
+
+    let tables = if !snapshot_config.collect_count_per_worker {
+        if leader {
+            all_tables
+        } else {
+            return Ok(None);
+        }
+    } else {
+        tables
+    };
+
+    if tables.is_empty() {
+        return Ok(None);
+    }
+
+    let task_name = format!("timely-{worker_id} PG snapshot counter");
+    // We create a new connection here so that postgres
+    let client = connection_config
+        .connect(&task_name, &connection_context.ssh_tunnel_manager)
+        .await?;
+
+    set_statement_timeout(&client, statement_timeout).await?;
+
+    // If we want a strict count, we want to count the rows in the snapshot
+    // determined in the operator.
+    if snapshot_config.collect_strict_count || snapshot_config.fallback_to_strict_count {
+        use_snapshot(&client, snapshot).await?
+    }
+
+    let jh = mz_ore::task::spawn(|| format!("pg_source_count"), async move {
+        let metrics = &metrics;
+        let client = &client;
+        let results = iter(tables)
+            .then(|(table_name, oid)| {
```

Comment:
In the beginning I thought we were doing some fancy concurrent stream processing thing but this will go over the tables serially so why are we not using a for loop?This is way too magical. It is possible to write it as normal for loops and I strongly recommend we do so. This is my attempt (I tried it to ensure it's possible before commenting):

```diff
diff --git a/src/storage/src/source/postgres/snapshot.rs b/src/storage/src/source/postgres/snapshot.rs
index e699ccbe4e..2a161f3ac4 100644
--- a/src/storage/src/source/postgres/snapshot.rs
+++ b/src/storage/src/source/postgres/snapshot.rs
@@ -138,9 +138,8 @@ use std::rc::Rc;
 use std::str::FromStr;
 use std::time::Duration;
 
+use anyhow::{bail, Context};
 use differential_dataflow::{AsCollection, Collection};
-use futures::future::FutureExt;
-use futures::stream::{iter, StreamExt};
 use futures::TryStreamExt;
 use mz_expr::MirScalarExpr;
 use mz_ore::result::ResultExt;
@@ -151,6 +150,7 @@ use mz_postgres_util::simple_query_opt;
 use mz_postgres_util::tunnel::Config;
 use mz_repr::{Datum, DatumVec, Diff, GlobalId, Row};
 use mz_sql_parser::ast::{display::AstDisplay, Ident};
+use mz_storage_types::parameters::PgSourceSnapshotConfig;
 use mz_storage_types::sources::{MzOffset, PostgresSourceConnection};
 use mz_timely_util::builder_async::{
     Event as AsyncEvent, OperatorBuilder as AsyncOperatorBuilder, PressOnDropButton,
@@ -626,92 +626,27 @@ async fn record_table_sizes(
     let jh = mz_ore::task::spawn(|| format!("pg_source_count"), async move {
         let metrics = &metrics;
         let client = &client;
-        let results = iter(tables)
-            .then(|(table_name, oid)| {
-                async move {
-                    let mut count_latency: f64 = 0.0;
-                    use mz_ore::metrics::MetricsFutureExt;
-
-                    let count_query = format!("SELECT count(*) as count from {}", table_name,);
-                    let count_query = simple_query_opt(client, &count_query)
-                        .map(|res| {
-                            let count = res?.ok_or_else(|| {
-                                anyhow::anyhow!(
-                                    "{}, failed to get count for {}",
-                                    source_id,
-                                    table_name
-                                )
-                            })?;
-                            let count: i64 = count.get("count").unwrap().parse().unwrap();
-                            Ok::<_, anyhow::Error>(count)
-                        })
-                        .wall_time()
-                        .set_at(&mut count_latency);
-
-                    let mut estimate_latency: f64 = 0.0;
-                    let estimate_query = format!(
-                        "SELECT reltuples::bigint AS estimate_count \
-                                FROM pg_class WHERE oid = '{}'",
-                        oid
-                    );
-                    let estimate_query = simple_query_opt(client, &estimate_query)
-                        .map(|res| {
-                            let estimate_count = res?.ok_or_else(|| {
-                                anyhow::anyhow!(
-                                    "{}, failed to get estimate count for {}",
-                                    source_id,
-                                    table_name
-                                )
-                            })?;
-                            let estimate_count: i64 = estimate_count
-                                .get("estimate_count")
-                                .unwrap()
-                                .parse()
-                                .unwrap();
-                            Ok::<_, anyhow::Error>(estimate_count)
-                        })
-                        .wall_time()
-                        .set_at(&mut estimate_latency);
-
-                    let (count, estimate_count) = if snapshot_config.collect_strict_count {
-                        (Some(count_query.await?), estimate_query.await?)
-                    } else {
-                        let estimate = estimate_query.await?;
-
-                        (
-                            // Postgres returns an estimate of -1 if the table doesn't have
-                            // sufficient writes/analysis/vacuuming happening.
-                            if estimate == -1 && snapshot_config.fallback_to_strict_count {
-                                Some(count_query.await?)
-                            } else {
-                                None
-                            },
-                            estimate,
-                        )
-                    };
-
-                    metrics.record_table_estimate(
-                        table_name.clone(),
-                        estimate_count,
-                        estimate_latency,
-                    );
-                    if let Some(count) = count {
-                        metrics.record_table_count(table_name.clone(), count, count_latency)
+
+        let mut result = Ok(());
+        for (table, oid) in tables {
+            match collect_table_statistics(client, snapshot_config, &table, oid).await {
+                Ok(stats) => {
+                    if let Some(count) = stats.estimate_count {
+                        metrics.record_table_estimate(table.clone(), count, stats.estimate_latency);
+                    }
+                    if let Some(count) = stats.count {
+                        metrics.record_table_count(table.clone(), count, stats.count_latency);
                     }
-                    Ok::<(), anyhow::Error>(())
                 }
-                .map(move |res| {
-                    if res.is_err() && !snapshot_config.wait_for_count {
-                        warn!(?res, "error when collecting pg count");
+                Err(err) => {
+                    if !snapshot_config.wait_for_count {
+                        warn!(?err, "error when collecting pg count");
                     }
-                    res
-                })
-            })
-            .collect::<Vec<_>>()
-            .await;
-
-        let result: Result<Vec<_>, _> = results.into_iter().collect();
-        result?;
+                    result = result.and(Err(err));
+                }
+            }
+        }
+        result.context(format!("{source_id}: "))?;
 
         // If we want a strict count, we want to count the rows in the snapshot
         // determined in the operator.
@@ -728,3 +663,51 @@ async fn record_table_sizes(
         Ok(Some(jh.abort_on_drop()))
     }
 }
+
+#[derive(Default)]
+struct TableStatistics {
+    count_latency: f64,
+    count: Option<i64>,
+    estimate_latency: f64,
+    estimate_count: Option<i64>,
+}
+
+async fn collect_table_statistics(
+    client: &Client,
+    config: PgSourceSnapshotConfig,
+    table: &str,
+    oid: u32,
+) -> Result<TableStatistics, anyhow::Error> {
+    use mz_ore::metrics::MetricsFutureExt;
+    let mut stats = TableStatistics::default();
+
+    let estimate_row = simple_query_opt(
+        client,
+        &format!("SELECT reltuples::bigint AS estimate_count FROM pg_class WHERE oid = '{oid}'"),
+    )
+    .wall_time()
+    .set_at(&mut stats.estimate_latency)
+    .await?;
+
+    match estimate_row {
+        Some(row) => match row.get("estimate_count").unwrap().parse().unwrap() {
+            -1 => stats.estimate_count = None,
+            n => stats.estimate_count = Some(n),
+        },
+        None => bail!("failed to get estimate count for {table}"),
+    }
+
+    // Postgres returns an estimate of -1 if the table doesn't have sufficient writes/analysis/vacuuming happening.
+    let should_fallback = config.fallback_to_strict_count && stats.estimate_count.is_none();
+    if config.collect_strict_count || should_fallback {
+        let count_row = simple_query_opt(client, &format!("SELECT count(*) as count from {table}"))
+            .wall_time()
+            .set_at(&mut stats.count_latency)
+            .await?;
+        match count_row {
+            Some(row) => stats.count = Some(row.get("count").unwrap().parse().unwrap()),
+            None => bail!("failed to get count for {table}"),
+        }
+    }
+    Ok(stats)
+}

```

Reply:
You are right, this can be a loop no!

Reply:
Thanks for making this entirely `git apply`-able!!